### ISSUE TITLE
EasyBuild installation test failed

### ISSUE THREAD

**Issue by JudyGardiner** [hpcugent/easybuild-framework#1539]
I'm fairly new to easybuild and wanted to evaluate the latest version for suitability on our system.  I followed your installation procedures successfully, but the unit test failed.  Here's the recommended troubleshooting information followed by the test output.  I would appreciate any insight you can provide.  --Judy

```
[oakley02]$ python -V
Python 2.6.6
[oakley02]$ type module
module is a function
module () 
{ 
    eval `$LMOD_CMD bash "$@"`
}
[oakley02]$ type -f module
-bash: type: module: not found
[oakley02]$ module --version

Modules based on Lua: Version 5.9.3  2015-04-04 06:40
    by Robert McLay mclay@tacc.utexas.edu

[oakley02]$ module av EasyBuild

------------------------------------------- /nfs/14/judithg/sciapps/eb/oak/modules/all -------------------------------------------
   EasyBuild/2.5.0

Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".

[oakley02]$ which -a eb
~/sciapps/eb/oak/software/EasyBuild/2.5.0/bin/eb
[oakley02]$ eb --version
This is EasyBuild 2.5.0 (framework: 2.5.0, easyblocks: 2.5.0) on host oakley02.osc.edu.
[oakley02]$ python -V
Python 2.6.6
[oakley02]$ type module
module is a function
module () 
{ 
    eval `$LMOD_CMD bash "$@"`
}
[oakley02]$ type -f module
-bash: type: module: not found
[oakley02]$ module --version

Modules based on Lua: Version 5.9.3  2015-04-04 06:40
    by Robert McLay mclay@tacc.utexas.edu

[oakley02]$ module av EasyBuild

------------------------------------------- /nfs/14/judithg/sciapps/eb/oak/modules/all -------------------------------------------
   EasyBuild/2.5.0

Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".

[oakley02]$ which -a eb
~/sciapps/eb/oak/software/EasyBuild/2.5.0/bin/eb
[oakley02]$ eb --version
This is EasyBuild 2.5.0 (framework: 2.5.0, easyblocks: 2.5.0) on host oakley02.osc.edu.

[oakley01]$ python -m test.framework.suite
WARNING: xmlrunner module not available, falling back to using unittest...

./nfs/14/judithg/sciapps/eb/oak/software/EasyBuild/2.5.0/lib/python2.6/site-packages/vsc_base-2.4.12-py2.6.egg/vsc/utils/testing.py:55: DeprecationWarning: BaseException.message has been deprecated as of Python 2.6
  elif hasattr(err, 'message'):
/nfs/14/judithg/sciapps/eb/oak/software/EasyBuild/2.5.0/lib/python2.6/site-packages/vsc_base-2.4.12-py2.6.egg/vsc/utils/testing.py:56: DeprecationWarning: BaseException.message has been deprecated as of Python 2.6
  msg = err.message
......................Skipping test_from_pr, no GitHub token available?
.Skipping test_from_pr, no GitHub token available?
................................(skipping GitRepository test)
.(skipping HgRepository test)
..(skipping SvnRepository test)
..............................Skipping test_dep_graph, since pygraph is not available
....Skipping test_dump_autopep8, since autopep8 is not available
................................................................F........................................Skipping test_from_pr, no GitHub token available?
..............................Skipping test_fetch_easyconfigs_from_pr, no GitHub token available?
..Skipping test_read, no GitHub token available?
.Skipping test_read_api, no GitHub token available?
.Skipping test_walk, no GitHub token available?
...............................................................................................................GC3Pie not available, skipping test
.............................
======================================================================
FAIL: Check whether ModulesTool instance is stateless between runs.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nfs/14/judithg/sciapps/eb/oak/software/EasyBuild/2.5.0/lib/python2.6/site-packages/easybuild_framework-2.5.0-py2.6.egg/test/framework/modules.py", line 364, in test_modules_tool_stateless
    self.assertErrorRegex(EasyBuildError, load_err_msg, modtool.load, ['GCC/4.6.3'])
  File "/nfs/14/judithg/sciapps/eb/oak/software/EasyBuild/2.5.0/lib/python2.6/site-packages/vsc_base-2.4.12-py2.6.egg/vsc/utils/testing.py", line 85, in assertErrorRegex
    self.assertTrue(regex.search(msg), "Pattern '%s' is found in '%s'" % (regex.pattern, msg))
AssertionError: Pattern 'cannot[\s
]*be[\s
]*loaded' is found in 'Changing environment as dictated by module failed: name 'false' is not defined (stdout: os.environ["LMOD_DEFAULT_MODULEPATH"] = "/tmp/eb-eQxp1x/eb-BOoV3p/eb-vxfYey/eb-vusnpt/eb-S72R_R/eb-h_3Zlw/eb-raVvkf/eb-SpKaE6/tmptWAYPB/modules/all:/tmp/eb-eQxp1x/eb-BOoV3p/eb-vxfYey/eb-vusnpt/eb-S72R_R/eb-h_3Zlw/eb-raVvkf/eb-SpKaE6/Core:/nfs/14/judithg/sciapps/eb/oak/software/EasyBuild/2.5.0/lib/python2.6/site-packages/easybuild_framework-2.5.0-py2.6.egg/test/framework/modules:/usr/local/share/lmodfiles/mpi/intel/15/mvapich2/2.1:/usr/local/share/lmodfiles/compiler/intel/15";
os.environ["MODULEPATH"] = "/tmp/eb-eQxp1x/eb-BOoV3p/eb-vxfYey/eb-vusnpt/eb-S72R_R/eb-h_3Zlw/eb-raVvkf/eb-SpKaE6/tmptWAYPB/modules/all:/tmp/eb-eQxp1x/eb-BOoV3p/eb-vxfYey/eb-vusnpt/eb-S72R_R/eb-h_3Zlw/eb-raVvkf/eb-SpKaE6/Core";
false
, stderr: Lmod has detected the following error: The following module(s) are unknown:
"GCC/4.6.3"

   Please check the spelling or version number. Also try "module spider ..."

)'

----------------------------------------------------------------------
Ran 372 tests in 3105.987s

FAILED (failures=1)
ERROR: Not all tests were successful.
Log available at /tmp/eb-eQxp1x/eb-BOoV3p/easybuild-tests-PY3lKv.log 
```

**Comment by boegel** [comment]
@JudyGardiner: this may be a bug in Lmod, rather than EasyBuild.

Can you try again after with an updated Lmod version, preferably with the latest from https://github.com/TACC/Lmod/releases (we know that the latest 6.0.1 available from http://sourceforge.net/projects/lmod/ has a(nother) bug you hit with EasyBuild too).

**Comment by JudyGardiner** [comment]
Thanks.  I’ll give it a try.  Will I need to reinstall EasyBuild or just rerun the test?

**Comment by boegel** [comment]
No need to reinstall, just update `$PATH` to point to the new `lmod` binary you've installed, and EasyBuild will use that.

**Comment by JudyGardiner** [comment]
I’m getting the same failure with the latest lmod.

[oakley02]$ module --version

Modules based on Lua: Version 6.0.24  2015-11-30 08:17
    by Robert McLay mclay@tacc.utexas.edu

**Comment by boegel** [comment]
Interesting... The specific test that is failing is expecting to see a failing `module load GCC/4.6.3`.

Loading the module *does* fail, but not as EasyBuild expects it. When using Lmod, it's expecting an error like `GCC/4.6.3 cannot be loaded`, while it's running into an error like `The following module(s) are unknown`.

I'm not sure why the error message differs, but that's basically the problem.

It's nothing to worry about though, EasyBuild usually doesn't parse the Lmod error message like this, it's something very specific to the test. I'll see if I can figure out why it happens, and then fix it.

Maybe @rtmclay has an idea?

**Comment by boegel** [comment]
OK, the problem seems to be that Lmod isn't seeing the `GCC/4.6.3` module at all, while the test is expecting the Lmod knows about the module's existence, but that it's not loadable because it's location is not in the current `$MODULEPATH`.

Looking at the test, it looks like the error you're seeing makes more sense, but somehow this has never popped up until now, despite this test not being new...

**Comment by boegel** [comment]
Seems like it's related to `$LMOD_DEFAULT_MODULEPATH`.

Do you have the module file `GCC/4.6.3` in `/nfs/14/judithg/sciapps/eb/oak/software/EasyBuild/2.5.0/lib/python2.6/site-packages/easybuild_framework-2.5.0-py2.6.egg/test/framework/modules` ?

In any case, it seems like we should make the test aware that the error message may differ, what Lmod prints exactly doesn't really matter, just as long as it can't load the module.

**Comment by JudyGardiner** [comment]
Thanks for your response.  If the error doesn’t signal future problems I’ll just ignore it.

We do use $LMOD_DEFAULT_MODULEPATH.  We need EasyBuild to coexist with our current software installations.
[oakley01]$ echo $LMOD_DEFAULT_MODULEPATH
/usr/local/share/lmodfiles/base

The GCC/4.6.3 module file exists in the location you specified.

Judy

**Comment by boegel** [comment]
Update: I can easily reproduce this problem if I'm using a sufficiently recent Lmod version.

I'm not seeing this with Lmod 5.6.3 (the minimal required version) or with 5.9(.0), but I am seeing it with Lmod 5.9.3 (which is what @JudyGardiner was using), and with 6.0.1.

```
$ TEST_EASYBUILD_MODULES_TOOL=Lmod python test/framework/modules.py
INFO: This is (based on) vsc.install.shared_setup 0.9.12
INFO: run_tests from base dir /Users/kehoste/work/easybuild-framework/test/framework (using executable /Users/kehoste/work/easybuild-framework/test/framework/modules.py)
Couldn't import dot_parser, loading of dot files will not be possible.
......F.....
======================================================================
FAIL: test_modules_tool_stateless (__main__.ModulesTest)
Check whether ModulesTool instance is stateless between runs.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/framework/modules.py", line 364, in test_modules_tool_stateless
    self.assertErrorRegex(EasyBuildError, load_err_msg, modtool.load, ['GCC/4.6.3'])
  File "/Users/kehoste/work/vsc-install/lib/vsc/install/testing.py", line 129, in assertErrorRegex
    self.assertTrue(regex.search(msg), "Pattern '%s' is found in '%s'" % (regex.pattern, msg))
AssertionError: Pattern 'cannot[\s
]*be[\s
]*loaded' is found in 'Changing environment as dictated by module failed: name 'false' is not defined (stdout: os.environ["LMOD_DEFAULT_MODULEPATH"] = "/private/var/folders/8s/_frgh9sj6m744mxt5w5lyztr0000gn/T/eb-or6plh/eb-PUcQsH/tmpqPFi1t/modules/all:/private/var/folders/8s/_frgh9sj6m744mxt5w5lyztr0000gn/T/eb-or6plh/eb-PUcQsH/Core:/Users/kehoste/work/easybuild-framework/test/framework/modules";
os.environ["MODULEPATH"] = "/private/var/folders/8s/_frgh9sj6m744mxt5w5lyztr0000gn/T/eb-or6plh/eb-PUcQsH/tmpqPFi1t/modules/all:/private/var/folders/8s/_frgh9sj6m744mxt5w5lyztr0000gn/T/eb-or6plh/eb-PUcQsH/Core:/var/folders/8s/_frgh9sj6m744mxt5w5lyztr0000gn/T/eb-or6plh/eb-PUcQsH/Core";
false
, stderr: Lmod has detected the following error: The following module(s) are unknown:
"GCC/4.6.3"

   Please check the spelling or version number. Also try "module spider ..."

)'

----------------------------------------------------------------------
Ran 12 tests in 76.339s

FAILED (failures=1)
```

**Comment by rtmclay** [comment]
Sites and users control MODULEPATH, Lmod controls LMOD_DEFAULT_MODULEPATH.  To know what is going on I'll need to see a "module -D load GCC/4.6.3 > load.log

**Comment by boegel** [comment]
OK, found the problem, `$LMOD_DEFAULT_MODULEPATH` was leaking into the test.

Fixed in https://github.com/hpcugent/easybuild-framework/pull/1570, I'll get this included in EasyBuild v2.6.0.

Thanks for the bug report @JudyGardiner!
